shadowsocks/udp.go

```diff
@@ -186,46 +186,79 @@ func (s *udpService) Stop() error {
 	return s.clientConn.Close()
 }
 
-type natentry struct {
-	conn net.PacketConn
+const (
+	modeFresh     = 0    // No packets have been sent.
+	modeSingleDNS = iota // One packet has been sent, a DNS query.
+	modeMultiDNS  = iota // Multiple packets have been sent, all DNS.
+	modeGeneral   = iota // Non-DNS packets have been sent.
+)
+
+type natconn struct {
+	net.PacketConn
 	// We store the client location in the NAT map to avoid recomputing it
 	// for every outbound packet in a UDP-based connection.
 	clientLocation string
+	// NAT timeout to apply for non-DNS connections.
+	defaultTimeout time.Duration
+	// Indicates whether DNS optimizations apply.
+	mode int32
+}
+
+func (c *natconn) WriteTo(buf []byte, dst net.Addr) (int, error) {
+	_, port, _ := net.SplitHostPort(dst.String())
+
+	// Update mode.
+	mode := c.mode
+	if port == "53" {
+		if mode == modeFresh {
+			mode = modeSingleDNS
+		} else if mode == modeSingleDNS {
+			mode = modeMultiDNS
+		}
+	} else {
+		mode = modeGeneral
+	}
+	atomic.StoreInt32(&c.mode, mode)
```

Comment:
I'm not sure you're gaining much from the atomic operation here. You still have a race condition where the new mode may be calculated using a stale value. I mean, WriteTo is not concurrency-safe.

Reply:
I was able to get rid of the atomics in this version.